metadata/src/main/java/org/apache/kafka/queue/KafkaEventQueue.java

```diff
@@ -0,0 +1,420 @@
+    }
+
+    private class EventHandler implements Runnable {
+        /**
+         * Event contexts indexed by tag.  Events without a tag are not included here.
+         */
+        private final Map<String, EventContext> tagToEventContext = new HashMap<>();
+
+        /**
+         * The head of the event queue.
+         */
+        private final EventContext head = new EventContext(null, null, null);
+
+        /**
+         * An ordered map of times in monotonic nanoseconds to events to time out.
+         */
+        private final TreeMap<Long, EventContext> delayMap = new TreeMap<>();
+
+        /**
+         * A condition variable for waking up the event handler thread.
+         */
+        private final Condition cond = lock.newCondition();
+
+        @Override
+        public void run() {
+            try {
+                handleEvents();
+                cleanupEvent.run();
+            } catch (Throwable e) {
+                log.warn("event handler thread exiting with exception", e);
+            }
+        }
+
+        private void remove(EventContext eventContext) {
+            eventContext.remove();
+            if (eventContext.deadlineNs != null) {
+                delayMap.remove(eventContext.deadlineNs);
+                eventContext.deadlineNs = null;
+            }
+            if (eventContext.tag != null) {
+                tagToEventContext.remove(eventContext.tag, eventContext);
+                eventContext.tag = null;
+            }
+        }
+
+        private void handleEvents() throws InterruptedException {
+            EventContext toTimeout = null;
+            EventContext toRun = null;
+            while (true) {
+                if (toTimeout != null) {
+                    toTimeout.completeWithTimeout();
+                    toTimeout = null;
+                } else if (toRun != null) {
+                    toRun.run();
+                    toRun = null;
+                }
+                lock.lock();
+                try {
+                    long awaitNs = Long.MAX_VALUE;
+                    Map.Entry<Long, EventContext> entry = delayMap.firstEntry();
+                    if (entry != null) {
+                        // Search for timed-out events or deferred events that are ready
+                        // to run.
+                        long now = time.nanoseconds();
+                        long timeoutNs = entry.getKey();
+                        EventContext eventContext = entry.getValue();
+                        if (timeoutNs <= now) {
+                            if (eventContext.insertionType == EventInsertionType.DEFERRED) {
+                                // The deferred event is ready to run.  Prepend it to the
+                                // queue.  (The value for deferred events is a schedule time
+                                // rather than a timeout.)
+                                remove(eventContext);
+                                head.insertAfter(eventContext);
```

Comment:
Unless I'm missing something, the event that got prepended to the queue will be immediately run on the next iteration of the loop.  There is no opportunity to prepend multiple deferred events to the toRun queue, since only one event can be prepended per loop run, and that event will be immediately run at the top of the loop.

Reply:
Hmm, after the event is prepended, we continue to the next while loop before dequeuing and setting toRun, right? Then, in the next loop, we could be prepending the next deferred item before the previous one runs.

Reply:
> Unless I'm missing something, the event that got prepended to the queue will be immediately run on the next iteration of the loop. There is no opportunity to prepend multiple deferred events to the toRun queue, since only one event can be prepended per loop run, and that event will be immediately run at the top of the loop.

Hmm, after the event is prepended, we continue to the next while loop before dequeuing and setting toRun, right? Then, in the next loop, we could be prepending the next deferred item before the previous one runs.

Reply:
There is a `continue;` at line 219 so that means that `toRun` is not set in this iteration of the loop did you mean to have this instead: `toRun = eventContext;`

Reply:
I looked at this again, and yes, you are right: it should just be `toRun = eventContext`.  Fixed.